### PR TITLE
Adding in-memory scheduler compensation job

### PR DESCRIPTION
**Pull request by robzienert**
Enables performing deployments or restarts with the in-memory scheduler so that it will execute any missed cron executions. Can be configured as below.

```
scheduler:
  enabled: true
  compensationJob:
    enabled: true
    windowMs: 30000 # optional
```

The compensation job will ignore any missed executions that fall below the configured time window.

@spinnaker/netflix-reviewers @duftler PTAL

### CODE REVIEW

File: echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+  transient Subscription subscription
+
+  @Autowired
+  MissedPipelineTriggerCompensationJob(Scheduler scheduler,
+                                       PipelineCache pipelineCache,
+                                       OrcaService orcaService,
+                                       PipelineInitiator pipelineInitiator,
+                                       CounterService counter,
+                                       @Value('${scheduler.compensationJob.windowMs:30000}') long compensationWindowMs,
```
Comment (ajordens) [inline echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy]: This is probably a little small for a default window?

Just to confirm the behavior ... we'll auto-kickoff a pipeline for any triggers that should have fired between `now` and `now - compensationWindowMs`?

In such scenarios we would want to err on the side of a larger window?

File: echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    List<String> ids = getPipelineConfigIds(pipelines, triggers)
+    orcaService.getLatestPipelineExecutions(ids)
+      .subscribe({
+        onOrcaResponse(it, pipelines, triggers)
+      }, { onOrcaError(it) })
+  }
+
+  void onOrcaResponse(Collection<PipelineResponse> response, List<Pipeline> pipelines, List<Trigger> triggers) {
+    triggers.each { trigger ->
+      Pipeline pipeline = pipelines.find { it.triggers.contains(trigger) }
```
Comment (ajordens) [inline echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy]: Might be safer to compare on `trigger.id` than the auto-generated `equals` ... probably just a personal preference of mine tho so don't feel obligated to change.

File: echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      def lastExecution = executions.first()
+      if (lastExecution == null) {
+        return
+      }
+
+      def expr = new CronExpression(trigger.cronExpression)
+      expr.timeZone = dateContext.timeZone
+
+      if (missedExecution(expr, lastExecution, dateContext.triggerWindowFloor, dateContext.now)) {
+        log.info("Triggering missed execution on pipeline ${pipeline}")
```
Comment (ajordens) [inline echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy]: do we need the whole pipeline logged, or is `application`/`pipelineConfigId` sufficient?

File: echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    def nextExecution = expr.getNextValidTimeAfter(lastExecution)
+    while (true) {
+      if (nextExecution.after(windowFloor)) {
+        break
+      }
+      nextExecution = expr.getNextValidTimeAfter(nextExecution)
+    }
+
+    return nextExecution.before(now)
```
Comment (ajordens) [inline echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy]: Maybe worth a comment either here or javadoc on the method indicating that we're looking for any that should have been fired between `windowFloor` and `now` (where `windowFloor` is `now - windowMs` right?)

File: echo-scheduler/src/test/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/MissedPipelineTriggerCompensationJobSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    def now = getDateOffset(nowMinutes)
+
+    when:
+    def result = MissedPipelineTriggerCompensationJob.missedExecution(expr, lastExecution, windowFloor, now)
+
+    then:
+    result == missedExecution
+
+    where:
+    cronExpression     | lastExecutionMinutes | windowFloorMinutes | nowMinutes | missedExecution
```
Comment (ajordens) [inline echo-scheduler/src/test/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/MissedPipelineTriggerCompensationJobSpec.groovy]: minor convention but I'd suggest using `||` as a delimiter between inputs and expected results ...

```
windowFloorMinutes | nowMinutes || missedExecution
```

File: echo-scheduler/src/test/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/MissedPipelineTriggerCompensationJobSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    '* 0/30 * * * ? *' | 0                    | 0                  | 0          | false
+    '* 0/30 * * * ? *' | 0                    | 30                 | 90         | false
```
Comment (ajordens) [inline echo-scheduler/src/test/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/MissedPipelineTriggerCompensationJobSpec.groovy]: I'd suggest adding some comments that describe what edge cases you're covering:

```
0 | 30 | 90 | false // blah blah blah
```

File: echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy (hunk 1 above)
Comment (robzienert) [inline echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy]: That's right - we'll kickoff anything between `now` and `now - compensationWindowMs`. Thanks for the call out here, I meant to go for 30 minutes, not 30 seconds. Not sure where my head was.

**Comment by ajordens** [comment]
30 minutes sounds perfect.

**Comment by robzienert** [comment]
PR comments addressed.

**Comment by ajordens** [comment]
LGTM

Once this goes live, where does this leave us with cassandra / scheduled actions? Running everything in-memory?

**Comment by robzienert** [comment]
Only as far as echo is concerned, we'd be able to deploy with in-memory scheduled actions.
